### PR TITLE
2.5D Occlusion and Testing Update

### PR DESCRIPTION
**Pull request by acpaquette**
Allows cam2map to account for occlusion when projecting images. Also updates cam2map with the application changes and test, and the application change for map2cam.

## Description
Allows cam2map to account for occluded portions of images when they are map projected. This option can only be used with the reverse patch algorithm and is turned off by default to keep any currently used pipelines for generating map projected products to remain the same.

This PR also updates cam2map and map2cam to be callable as functions. The tests for cam2map where also updated with additional fixtures as well as a few new mocks.

## Related Issue
This is related to a larger occlusion project with no specific issues on the ISIS repo.

## Motivation and Context
This allows ISIS to more accurately project highly oblique images, so long as the image is accurately coregistered to its associated DEM.

## How Has This Been Tested?
This has been tested with the current ISIS tests along with a new set of gtests.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation change (update to the documentation; no code change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have read and agree to abide by the [Code of Conduct](https://usgs-astrogeology.github.io/code/)
- [x] I have read the [**CONTRIBUTING**](https://github.com/USGS-Astrogeology/ISIS3/blob/dev/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

## Licensing
This project is mostly composed of free and unencumbered software released into the public domain, and we are unlikely to accept contributions that are not also released into the public domain. Somewhere near the top of each file should have these words:

> This work is free and unencumbered software released into the public domain. In jurisdictions that recognize copyright laws, the author or authors of this software dedicate any and all copyright interest in the software to the public domain.

- [x] I dedicate any and all copyright interest in this software to the public domain. I make this dedication for the benefit of the public at large and to the detriment of my heirs and successors. I intend this dedication to be an overt act of relinquishment in perpetuity of all present and future rights to this software under copyright law.

### CODE REVIEW

File: isis/src/base/apps/cam2map/cam2map.cpp [hunk 1]
```diff
@@ -0,0 +1,565 @@
...
+      transform = new cam2mapReverse(icube->sampleCount(),
+                                     icube->lineCount(), incam, samples,lines,
+                                     outmap, trim, occlusion);
+
+      int patchSize = ui.GetInteger("PATCHSIZE");
+      int minPatchSize = 4;
+      if (patchSize < minPatchSize) {
+        patchSize = minPatchSize;
+      }
+      p.SetTiling(patchSize, patchSize);
```
Comment (acpaquette) [inline isis/src/base/apps/cam2map/cam2map.cpp:334]: Originally:
```
int minPatchSize = 4;
if (patchSize < minPatchSize) {
    minPatchSize = patchSize;
}
p.SetTiling(patchSize, minPatchSize);
```

File: isis/src/base/apps/cam2map/cam2map.cpp [hunk 2]
```diff
@@ -0,0 +1,565 @@
...
+    PvlGroup &userGrp = userMap.findGroup("Mapping", Pvl::Traverse);
+
+    cam2map(&icube, userMap, userGrp, ui, log);
+  }
+
+
+  void cam2map(Cube *icube, Pvl &userMap, PvlGroup &userGrp, UserInterface &ui, Pvl *log) {
+    ProcessRubberSheet p;
+    cam2map(icube, userMap, userGrp, p, ui, log);
+  }
```
Comment (acpaquette) [inline isis/src/base/apps/cam2map/cam2map.cpp:37]: New method so a mocked process object can be injected into the test.

**Review by acpaquette — COMMENTED** [review]
There are two notable changes in cam2map that I have outlined so people can discuss them if necessary. The rest of cam2map should be the same as what was previously in the cam2map/main.cpp

File: isis/src/base/apps/cam2map/cam2map.cpp [hunk 3]
```diff
@@ -0,0 +1,565 @@
...
+    inLine = p_incam->Line();
+
+    // Good to ground one last time to check for occlusion
+    p_incam->SetImage(inSample, inLine);
+
+    if (p_occlusion){
+      if (abs(lat - p_incam->UniversalLatitude()) > 0.00001 || abs(lon - p_incam->UniversalLongitude()) > 0.00001) {
+        return false;
+      }
+    }
```
Comment (acpaquette) [inline isis/src/base/apps/cam2map/cam2map.cpp:549]: Also the actual occlusion handling in the cam2mapReverse Xform

**Review by scsides — APPROVED** [review]

**Pull request merged**